wallet/src/bin/random_wallet.rs

```diff
@@ -226,75 +236,142 @@ async fn main() {
-                String::from("the native asset")
-            } else if *asset == my_asset.code {
-                String::from("my asset")
-            } else {
-                asset.to_string()
-            },
-            recipient,
-        );
-        let txn = match wallet
-            .transfer(Some(&address), asset, &[(recipient.address(), amount)], fee)
-            .await
-        {
-            Ok(txn) => txn,
-            Err(err) => {
-                event!(Level::ERROR, "Error generating transfer: {}", err);
-                continue;
+                event!(
+                    Level::INFO,
+                    "transferring {} units of {} to user {}",
+                    amount,
+                    if *asset == AssetCode::native() {
+                        String::from("the native asset")
+                    } else if *asset == my_asset.code {
+                        rec_addresses.push(recipient.clone());
+                        String::from("my asset")
+                    } else {
+                        asset.to_string()
+                    },
+                    recipient,
+                );
+                let txn = match wallet
+                    .transfer(Some(&address), asset, &[(recipient.address(), amount)], fee)
+                    .await
+                {
+                    Ok(txn) => txn,
+                    Err(err) => {
+                        event!(Level::ERROR, "Error generating transfer: {}", err);
+                        continue;
+                    }
+                };
+                match wallet.await_transaction(&txn).await {
+                    Ok(status) => {
+                        if !status.succeeded() {
+                            // Transfers are allowed to fail. It can happen, for instance, if we get starved
+                            // out until our transfer becomes too old for the validators. Thus we make this
+                            // a warning, not an error.
+                            event!(Level::WARN, "transfer failed!");
+                        }
+                    }
+                    Err(err) => {
+                        event!(Level::ERROR, "error while waiting for transaction: {}", err);
+                    }
+                }
             }
-        };
-        match wallet.await_transaction(&txn).await {
-            Ok(status) => {
-                if !status.succeeded() {
-                    // Transfers are allowed to fail. It can happen, for instance, if we get starved
-                    // out until our transfer becomes too old for the validators. Thus we make this
-                    // a warning, not an error.
-                    event!(Level::WARN, "transfer failed!");
+            OperationType::Freeze => {
+                // Choose a random person who we sent our asset at one point
+                // This may fail if they've transfered that asset
+                if rec_addresses.is_empty() {
```

Comment:
If you audit, you can actually just use this wallet to list records owned by other wallets that this wallet can freeze (using the `records` iterator with some filtering)